arangod/Aql/Functions.cpp

```diff
@@ -2779,6 +2779,324 @@ AqlValue Functions::IsInPolygon(arangodb::aql::Query* query,
+    RegisterWarning(query, "GEO_POLYGON",
+                    TRI_ERROR_QUERY_ARRAY_EXPECTED);
+    return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
+  }
+
+  VPackBuilder b;
+
+  b.add(VPackValue(VPackValueType::Object));
+  b.add("type", VPackValue("Polygon"));
+  b.add("coordinates", VPackValue(VPackValueType::Array));
+
+  AqlValueMaterializer materializer(trx);
+  VPackSlice s = materializer.slice(geoArray, false);
+  for (auto const& v : VPackArrayIterator(s)) {
+    if (v.isArray() && v.length() > 2) {
+      b.openArray();
+      for (auto const& coord : VPackArrayIterator(v)) {
+        if (coord.isNumber()) {
+          b.add(VPackValue(coord.getNumber<double>()));
+        } else if (coord.isArray()) {
+          if (coord.length() < 2) {
+            RegisterWarning(query, "GEO_POLYGON", Result(
+                  TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+                  "a polygon needs at least three positions"));
+            return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
+          } else {
+            b.openArray();
+            for (auto const& innercord : VPackArrayIterator(coord)) {
+              if (innercord.isNumber()) {
+                b.add(VPackValue(innercord.getNumber<double>()));
+              } else if (innercord.isArray()) {
+                if (innercord.at(0).isDouble() &&innercord.at(1).isDouble()) {
+                  b.openArray();
+                  b.add(VPackValue(innercord.at(0).getDouble()));
+                  b.add(VPackValue(innercord.at(1).getDouble()));
+                  b.close();
+                } else {
+                  RegisterWarning(query, "GEO_POLYGON", Result(
+                        TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+                        "not an array"));
+                  return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
+                }
+              } else {
+                RegisterWarning(query, "GEO_POLYGON", Result(
+                      TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+                      "not a numeric value"));
+                return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
+              }
+            }
+            b.close();
+          }
+        } else {
+          RegisterWarning(query, "GEO_POLYGON", Result(
+                TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+                "not an array containing positions"));
+          return AqlValue(arangodb::basics::VelocyPackHelper::NullValue());
+        }
+      }
+      b.close();
+    } else if (v.isArray() && v.length() == 2) {
+        if (s.length() > 2) {
+        b.openArray();
+        for (auto const& innercord : VPackArrayIterator(v)) {
+          if (innercord.isNumber()) {
+            b.add(VPackValue(innercord.getNumber<double>()));
+          } else if (innercord.isArray()) {
+            if (innercord.at(0).isDouble() &&innercord.at(1).isDouble()) {
```

Comment:
Can we better check for `isNumber()` here? End users may well use integers and not only doubles.